### DISCUSSION TITLE
How to backup on MacOS Docker?

### DISCUSSION THREAD

**Discussion by FuadEfendi** [appwrite/appwrite#4280] [Q&A] [answered]
Hello,

I am trying to setup DEV environment on MacOS and I need backup/restore; so far I only found this useful script to get Linux VM shell:
```
docker run -it --rm --privileged --pid=host justincormack/nsenter1
```
Appwrite is installed in `/var/lib/docker/overlay2/`

I found some examples how to do backup, but those require "navigate to the Appwrite directory" ??? 
https://gist.github.com/stnguyen90/fee636ff652b8ecbf761935b2aa254fb

Please help

**Comment by stnguyen90** [comment, marked as answer]
Duplicate of https://github.com/appwrite/appwrite/discussions/3722

**Reply by FuadEfendi** [reply]
It is not duplicate of #3722 because this question is specific to **MacOS Docker Application.**

And I found the answer: use *manual* installation as described in https://appwrite.io/docs/installation#manual

Create folder, place these two configuration files, "manually install", and after that you can run backup/restore scripts from this folder. 

It should be highly advocated to prefer *manual* installation in the documentation, and not only on MacOS.

**Reply by stnguyen90** [reply]
@FuadEfendi I'm not sure how using MacOS Docker makes a difference. 🧐 If you installed Appwrite as our installation guide instructs, you should have an appwrite folder with the docker-compose.yml and .env files.

**Reply by FuadEfendi** [reply]
>  I'm not sure how using MacOS Docker makes a difference.

MacOS Docker is powered by Linux VM; "docker" is run inside "Linux VM" on top of MacOS. There is no "docker-compose.yml" files created neither on MacOS nor in this Linux VM (at least, it is not obvious what to do in this VM and how to backup)

Here is where "appwrite" components are located in Linux VM (run by MacOS Docker Application):
```
/var/lib/docker # ls image
overlay2
/var/lib/docker # find . -name appwrite
./overlay2/44478ad3e48270dd70c71cc24b12602a1475374c1d60afa9efea530ce9dc6a71/merged/usr/src/code/vendor/appwrite
./overlay2/4281480d59b68782362b25a0c72ca27a2653295f3a5d8ed5390fd2cfb5f411f6/merged/usr/src/code/vendor/appwrite
./overlay2/661175c13e375cd87d31897b99db1793cfbc4335f3482aeb351c36c35db15436/merged/usr/src/code/vendor/appwrite
./overlay2/979d5a1cff540ca64db34977f7a20c6d5e8ea453e40a1ae6d6ac948a3c206d54/merged/usr/src/code/vendor/appwrite
./overlay2/78e43383048a0b9843fb5cf24ce5ee8861b88aad7d2540c2e0163847a60454d5/merged/usr/src/code/vendor/appwrite
./overlay2/525bd6809888f5bde6ca322a01e3f8c92d25485a7b0104be16e34f78716e76dc/merged/usr/src/code/vendor/appwrite
./overlay2/d6e0705dd7226575eb2a84861be6b1112928566e837df75bc8466ba73c75fd5c/merged/usr/src/code/vendor/appwrite
./overlay2/a561cdb692df8c20bb94f76730e1dd68374833e6f2f0f7681b387026024b815d/merged/usr/src/code/vendor/appwrite
./overlay2/4211fe75cda3f5cd181cd35536d04ba27656e5dfbaede5ad7a9d5c3b0a13bdfd/merged/usr/src/code/vendor/appwrite
./overlay2/93892bf2cb6af134e38a39db9da8aad479bb509cb2397fba8d91308dbfe84faf/diff/usr/src/code/vendor/appwrite
./overlay2/27c7e5ec2152d6b189300f8368ef4bda8c2aab1b5381948d3f4794a80a9fbf5f/merged/usr/src/code/vendor/appwrite
./overlay2/da78c390c4872dda375e2e63bf9241398638e097fe05871a4d78331e221db6af/merged/usr/src/code/vendor/appwrite
./overlay2/2e01823feddc02983518eee24057e43732ee26544d166bd17165cede2ed0534d/merged/usr/src/code/vendor/appwrite
./overlay2/2710507cd72ff33b1c61c5d2f779e0d9b6e958f4089ed3898aefc4d0bbc8e156/merged/usr/src/code/vendor/appwrite
./overlay2/0894dbdb4829eaf655f3d60f0cd273e7ff9640c625434ec4aa038d7c9192daad/merged/usr/src/code/vendor/appwrite
./overlay2/f272e4cdd0e05e7d6a44b3fcee2f68d0c036cbd8067be8ae802c79eaab588897/merged/usr/src/code/vendor/appwrite
./overlay2/58f5eb100f453cb0a98f23a13a91358fb90c0bff90e31d689b6008c00a74b737/merged/usr/src/code/vendor/appwrite
./volumes/appwrite_appwrite-mariadb/_data/appwrite
/var/lib/docker # cd /usr/src/code
```

Does it look standard Linux environment? Not for me. I spend many hours before starting to suspect that I need physical file "docker-compose.yml" in the folder where I execute backup/restore script, and after that I found that it is trivial "manual" installation.

**Reply by stnguyen90** [reply]
@FuadEfendi, how did you install and start appwrite?

**Reply by FuadEfendi** [reply]
> how did you install and start appwrite?

At first, I followed this: https://appwrite.io/docs/installation#installWithDocker
```
docker run -it --rm \
    --volume /var/run/docker.sock:/var/run/docker.sock \
    --volume "$(pwd)"/appwrite:/usr/src/code/appwrite:rw \
    --entrypoint="install" \
    appwrite/appwrite:1.0.1
```
- and I had hard time to find "appwrite" installation/configuration folder.

Then, I tried *manual installation*, https://appwrite.io/docs/installation#manual
   1. create older `appwrite`
   2. download https://appwrite.io/install/compose
   3. download https://appwrite.io/install/env
   4. run command `docker compose up -d --remove-orphans`
   
   After that, I was able to run backup script from this folder.

**Reply by stnguyen90** [reply]
@FuadEfendi, so the [install command](https://appwrite.io/docs/installation#installWithDocker) you ran should have created an appwrite folder with the `docker-compose.yml` and `.env` file.

**Reply by FuadEfendi** [reply]
> /usr/src/code/appwrite



> @FuadEfendi, so where you ran that [install command](https://appwrite.io/docs/installation#installWithDocker), there should be an appwrite folder with the `docker-compose.yml` and `.env` file.

I used "Spotlight Search" to find all `docker-compose.yml` files created today, and I found only one which I explicitly downloaded. I assume other one was created (repeating) in Linux VM, `volume "$(pwd)"/appwrite:/usr/src/code/appwrite:rw`, but I could not find it.

Docker on MacOS has very special architecture, it cannot natively run containers, all containers are run in Linux VM, this is command to get terminal session (this one is simplest):
```
docker run -it --rm --privileged --pid=host justincormack/nsenter1
```

After that, I couldn't find any '/usr/src/code' in it; this VM probably was not designed for such filesystem surfing.

**Reply by stnguyen90** [reply]
@FuadEfendi, perhaps your spotlight didn't index the file? 🧐 Here's the output of my installation on my Mac:

<img width="806" alt="image" src="https://user-images.githubusercontent.com/1477010/193691715-46034a0c-36c3-44a8-8313-4cce744bf7a9.png">

**Reply by FuadEfendi** [reply]
Thank you for trying to reproduce!

I tried too, it worked this time. And I found `appwrite` folder created 9 hours earlier today (by guessing "where I was in the morning?"), somehow Spotlight doesn't index system-created files; manually downloaded one was instantly indexed.

Difficult day for me ;) almost 9-hours for troubleshooting. I am very sorry I took your time too. I was just guessing this folder would be somewhere in standard places, such as '~/.docker', never thought it will be subfolder at the script execution directory.

**Reply by stnguyen90** [reply]
Mystery solved! 🎉  Ya...I don't trust spotlight's file search 😆

**Comment by tripolskypetr** [comment]
If you facing the 500 Internal Server Error while using AppWrite Cloud data migration, [that json-per-document migration tool](https://github.com/react-declarative/appwrite-backup-tool) will be quite useful
